### ISSUE TITLE
tcpdump fails to build when openssl-1.1 was built with --api=1.1.0

### ISSUE THREAD

**Issue by Polynomial-C** [the-tcpdump-group/tcpdump#539]
This happens even with latest git head:

```
x86_64-pc-linux-gnu-gcc -ffloat-store -march=native -mtune=native -pipe -O2 -DHAVE_CONFIG_H  -DHAVE_CAP_NG_H  -D_U_="__attribute__((unused))" -I. -I/usr/include  -I/usr/include -march=native -mtune=native -pipe -O2 -Wl,-O1 -Wl,--hash-style=gnu -Wl,--sort-common -Wl,--as-needed -o tcpdump setsignal.o tcpdump.o version.o addrtoname.o addrtostr.o af.o ascii_strcasecmp.o checksum.o cpack.o gmpls.o gmt2local.o in_cksum.o ipproto.o l2vpn.o machdep.o nlpid.o oui.o parsenfsfh.o print.o print-802_11.o print-802_15_4.o print-ah.o print-ahcp.o print-aodv.o print-aoe.o print-ap1394.o print-arcnet.o print-arp.o print-ascii.o print-atalk.o print-atm.o print-babel.o print-beep.o print-bfd.o print-bgp.o print-bootp.o print-bt.o print-calm-fast.o print-carp.o print-cdp.o print-cfm.o print-chdlc.o print-cip.o print-cnfp.o print-dccp.o print-decnet.o print-dhcp6.o print-domain.o print-dtp.o print-dvmrp.o print-eap.o print-egp.o print-eigrp.o print-enc.o print-esp.o print-ether.o print-fddi.o print-forces.o print-fr.o print-frag6.o print-ftp.o print-geneve.o print-geonet.o print-gre.o print-hncp.o print-hsrp.o print-http.o print-icmp.o print-icmp6.o print-igmp.o print-igrp.o print-ip.o print-ip6.o print-ip6opts.o print-ipcomp.o print-ipfc.o print-ipnet.o print-ipx.o print-isakmp.o print-isoclns.o print-juniper.o print-krb.o print-l2tp.o print-lane.o print-ldp.o print-lisp.o print-llc.o print-lldp.o print-lmp.o print-loopback.o print-lspping.o print-lwapp.o print-lwres.o print-m3ua.o print-medsa.o print-mobile.o print-mobility.o print-mpcp.o print-mpls.o print-mptcp.o print-msdp.o print-msnlb.o print-nflog.o print-nfs.o print-nsh.o print-ntp.o print-null.o print-olsr.o print-openflow-1.0.o print-openflow.o print-ospf.o print-ospf6.o print-otv.o print-pgm.o print-pim.o print-pktap.o print-ppi.o print-ppp.o print-pppoe.o print-pptp.o print-radius.o print-raw.o print-resp.o print-rip.o print-ripng.o print-rpki-rtr.o print-rrcp.o print-rsvp.o print-rt6.o print-rtsp.o print-rx.o print-sctp.o print-sflow.o print-sip.o print-sl.o print-sll.o print-slow.o print-smtp.o print-snmp.o print-stp.o print-sunatm.o print-sunrpc.o print-symantec.o print-syslog.o print-tcp.o print-telnet.o print-tftp.o print-timed.o print-tipc.o print-token.o print-udld.o print-udp.o print-usb.o print-vjc.o print-vqp.o print-vrrp.o print-vtp.o print-vxlan.o print-vxlan-gpe.o print-wb.o print-zephyr.o print-zeromq.o netdissect.o signature.o strtoaddr.o util-print.o  strlcat.o strlcpy.o -lcap-ng -lcrypto -L/usr/lib -L/usr/lib64 -Wl,-rpath,/usr/lib64 -lpcap -lcap-ng
print-esp.o: In function `esp_print_decodesecret':
print-esp.c:(.text+0xb3b): undefined reference to `OpenSSL_add_all_algorithms'
collect2: error: ld returned 1 exit status
make: *** [Makefile:369: tcpdump] Error 1
```

**Comment by fxlb** [comment]
Is `libssl-dev` installed ?

**Comment by Polynomial-C** [comment]
Sorry for not being more verbose.

This is a Gentoo development machine with full installation of openssl-1.1.0 with all required libs and headers.
openssl-1.1 has a new API whose differences can be found [here](https://wiki.openssl.org/index.php/1.1_API_Changes).

**Comment by fxlb** [comment]
Nothing in 'API Changes' about `OpenSSL_add_all_algorithms` ...
What is the output of `./configure` ?

**Comment by Polynomial-C** [comment]
```
./configure --prefix=/usr --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --docdir=/usr/share/doc/tcpdump-9999 --htmldir=/usr/share/doc/tcpdump-9999/html --libdir=/usr/lib64 --disable-smb --with-chroot= --without-smi --with-crypto=/usr --with-user=tcpdump
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for x86_64-pc-linux-gnu-gcc... x86_64-pc-linux-gnu-gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether x86_64-pc-linux-gnu-gcc accepts -g... yes
checking for x86_64-pc-linux-gnu-gcc option to accept ISO C89... none needed
checking whether the compiler supports the -ffloat-store option... yes
checking for inline... inline
checking for __attribute__... yes
checking whether __attribute__((unused)) can be used without warnings... yes
checking whether __attribute__((noreturn)) can be applied to function pointers without warnings... yes
checking whether __attribute__((format)) can be used without warnings... yes
checking whether __attribute__((format)) can be applied to function pointers... yes
checking how to run the C preprocessor... x86_64-pc-linux-gnu-gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking rpc/rpc.h usability... yes
checking rpc/rpc.h presence... yes
checking for rpc/rpc.h... yes
checking rpc/rpcent.h usability... no
checking rpc/rpcent.h presence... no
checking for rpc/rpcent.h... no
checking netdnet/dnetdb.h usability... no
checking netdnet/dnetdb.h presence... no
checking for netdnet/dnetdb.h... no
checking for net/pfvar.h... no
checking for netinet/if_ether.h... yes
checking whether time.h and sys/time.h may both be included... yes
checking whether to enable the possibly-buggy SMB printer... no
checking whether to drop root privileges by default... to "tcpdump"
checking whether to chroot... no
checking for cap_enter... no
checking for cap_rights_limit... no
checking for cap_ioctls_limit... no
checking for openat... yes
checking whether to sandbox using capsicum... no
checking for library containing gethostbyname... none required
checking for library containing socket... none required
checking for library containing putmsg... none required
checking whether the operating system supports IPv6... yes
checking ipv6 stack type... linux-glibc
checking for dnet_htoa declaration in netdnet/dnetdb.h... no
checking for vfprintf... yes
checking for strlcat... no
checking for strlcpy... no
checking for strdup... yes
checking for strsep... yes
checking for getopt_long... yes
checking for fork... yes
checking for vfork... yes
checking for strftime... yes
checking for setlinebuf... yes
checking for alarm... yes
checking for vsnprintf... yes
checking for snprintf... yes
checking return type of signal handlers... void
checking for sigaction... yes
checking for library containing dnet_htoa... no
checking for main in -lrpc... no
checking for library containing getrpcbynumber... none required
checking for local pcap library... not found
checking for x86_64-pc-linux-gnu-pcap-config... no
checking for pcap-config... /usr/bin/pcap-config
checking for pcap_loop... yes
checking for ether_ntohost... yes
checking for buggy ether_ntohost... no
checking whether ether_ntohost is declared... no
checking netinet/ether.h usability... yes
checking netinet/ether.h presence... yes
checking for netinet/ether.h... yes
checking whether ether_ntohost is declared... yes
checking for dlpi_walk in -ldlpi... no
checking for pcap_list_datalinks... yes
checking for pcap_free_datalinks... yes
checking for pcap_set_datalink... yes
checking for pcap_datalink_name_to_val... yes
checking for pcap_datalink_val_to_description... yes
checking for pcap_breakloop... yes
checking for pcap_dump_ftell... yes
checking for pcap_create... yes
checking for pcap_set_tstamp_type... yes
checking for pcap_set_tstamp_precision... yes
checking for pcap_findalldevs... yes
checking for pcap_dump_flush... yes
checking for pcap_lib_version... yes
checking for pcap_setdirection... yes
checking for pcap_set_immediate_mode... yes
checking for pcap_if_t... yes
checking for pcap_set_parser_debug... no
checking whether pcap_debug is defined by libpcap... no
checking whether yydebug is defined by libpcap... no
checking for pcap_set_optimizer_debug... no
checking for bpf_dump... yes
checking for int8_t... yes
checking for int16_t... yes
checking for int32_t... yes
checking for int64_t... yes
checking for uint8_t... yes
checking for uint16_t... yes
checking for uint32_t... yes
checking for uint64_t... yes
checking for uintptr_t... yes
checking for u_int8_t... yes
checking for u_int16_t... yes
checking for u_int32_t... yes
checking for u_int64_t... yes
checking for inttypes.h... (cached) yes
checking whether inttypes.h defines the PRI[doxu]64 macros... yes
checking for pcap/bluetooth.h... yes
checking for pcap/nflog.h... yes
checking for pcap/usb.h... yes
checking for x86_64-pc-linux-gnu-ranlib... x86_64-pc-linux-gnu-ranlib
checking for x86_64-pc-linux-gnu-ar... x86_64-pc-linux-gnu-ar
checking if sockaddr struct has the sa_len member... no
checking if unaligned accesses fail... no
checking whether to use OpenSSL/libressl libcrypto... yes, using the version installed in /usr
checking openssl/crypto.h usability... yes
checking openssl/crypto.h presence... yes
checking for openssl/crypto.h... yes
checking for DES_cbc_encrypt in -lcrypto... yes
checking openssl/evp.h usability... yes
checking openssl/evp.h presence... yes
checking for openssl/evp.h... yes
checking for EVP_CIPHER_CTX_new... yes
checking whether to use libcap-ng... yes, if available
checking for capng_change_id in -lcap-ng... yes
checking cap-ng.h usability... yes
checking cap-ng.h presence... yes
checking for cap-ng.h... yes
checking for a BSD-compatible install... /usr/bin/install -c
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tcpdump.1
config.status: creating config.h
config.status: executing default-1 commands
```

**Comment by fxlb** [comment]
What is the `LIBS = ` line in Makefile ?

**Comment by fxlb** [comment]
Any change if you omit `--with-crypto=/usr` with `./configure` ?

**Comment by Polynomial-C** [comment]
Omitting ```--with-crypto=/usr``` with ```./configure``` yields the same compile error during make.

The LIBS variable looks like this:
```
# grep -F "LIBS =" Makefile
LIBS = -lcap-ng -lcrypto -L/usr/lib64 -Wl,-rpath,/usr/lib64 -lpcap
```

**Comment by Polynomial-C** [comment]
Oh, the above LIBS is from ```./configure``` *without* ```--with-crypto=/usr```.
Once I re-add that option LIBS looks like this:

```
# grep -F "LIBS =" Makefile
-lcap-ng -lcrypto -L/usr/lib -L/usr/lib64 -Wl,-rpath,/usr/lib64 -lpcap -lcap-ng
```

**Comment by fxlb** [comment]
First problem : twice `-lcap-ng`.
Any change if configure is only use with `--disable-smb --without-smi --with-user=tcpdump` ?

**Comment by Polynomial-C** [comment]
The build also fails on Debian although with a slightly differrent error message: 
https://breakpoint.cc/openssl-1.1-rebuild-2016-08-26/failed/tcpdump_4.7.4-1_amd64-2016-08-26T19%3A52%3A30Z

**Comment by orebokech** [comment]
But that's 4.7.4, which doesn't have the fix (b1f305da3dd).

Latest Git builds fine with OpenSSL 1.1.x on Debian, however the test suite fails because the resulting tcpdump segfaults when you use -E to decrypt.

**Comment by Polynomial-C** [comment]
Alright, this might be a problem with Gentoo's openssl package. I gonna close this issue and investigate on our side.
Thanks for your time and sorry that I have wasted so much of it.

**Comment by Polynomial-C** [comment]
Hi,

I've identified the issue. This linker error comes from openssl-1.1 being compiled with --api=1.1.0

This leads to `/usr/include/openssl/evp.h` not defining `OpenSSL_add_all_algorithms` anymore.

See https://github.com/openssl/openssl/blob/master/include/openssl/evp.h#L841 lines 841 to 867

I've reopened this issue in case you want to address it.

If you think this is not something that requires fixing may I suggest that you document the requirement of openssl-1.1.0 being compiled with --api=1.0.0 or --api=0.9.8 somewhere?

**Comment by mcr** [comment]
Lars Wendler <notifications@github.com> wrote:
    > If you think this is not something that requires fixing may I suggest that
    > you document the requirement of openssl-1.1.0 being compiled with --api=1.0.0
    > or --api=0.9.8 somewhere?

How are various distros compiling?  I would imagine either having two
packages, or can it be compiled with both APIs?

--
]               Never tell me the odds!                 | ipv6 mesh networks [
]   Michael Richardson, Sandelman Software Works        | network architect  [
]     mcr@sandelman.ca  http://www.sandelman.ca/        |   ruby on rails    [

**Comment by Polynomial-C** [comment]
Unfortunately it cannot compiled with both APIs.
I can only speak for Gentoo regarding the distros question. We have not yet decided what api we will use for openssl-1.1 but we won't be able to provide more than one API.

**Comment by guyharris** [comment]
Fixed in ad7df8bd34fa495cd9066ca138384b5d85b85523.
